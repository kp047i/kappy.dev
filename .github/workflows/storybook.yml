name: Storybook Tests

on:
  push:
  pull_request:

jobs:
  storybook:
    name: Storybook Tests
    runs-on: ubuntu-latest
    timeout-minutes: 60
    env:
      PLAYWRIGHT_BROWSERS_PATH: 0
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 22.13.0
      - name: Enable Corepack
        run: corepack enable
      - name: Install pnpm
        run: |
          corepack prepare pnpm@8.15.5 --activate
          pnpm --version
          which pnpm
      - name: Get pnpm store directory
        shell: bash
        run: |
          echo "STORE_PATH=$(pnpm store path --silent)" >> $GITHUB_ENV
      - name: Setup pnpm cache
        uses: actions/cache@v4
        with:
          path: ${{ env.STORE_PATH }}
          key: ${{ runner.os }}-pnpm-store-${{ hashFiles('**/pnpm-lock.yaml') }}
          restore-keys: |
            ${{ runner.os }}-pnpm-store-
      - name: Install dependencies
        run: pnpm install --frozen-lockfile
      - name: Install Playwright browsers
        run: pnpm exec playwright install --with-deps chromium chromium-headless-shell
      - name: Build Storybook
        run: pnpm build-storybook --quiet
      - name: Storybook Vitest tests
        env:
          CI: true
        run: pnpm run test:storybook
