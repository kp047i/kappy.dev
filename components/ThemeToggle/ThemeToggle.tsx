"use client";

import clsx from "clsx";
import { useTheme } from "next-themes";
import type { ReactNode } from "react";
import { FiMoon, FiSun } from "react-icons/fi";

const BUTTON_BASE =
  "flex h-9 w-9 items-center justify-center rounded-md border border-transparent text-sm transition-colors duration-200";

export function ThemeToggle() {
  const { theme, setTheme } = useTheme();

  return (
    <div className="flex gap-1 items-center p-1 rounded-lg border shadow-sm backdrop-blur border-secondary-100/80 bg-overlay-100/90 dark:border-base-800/60 dark:bg-base-800/60">
      <ThemeToggleButton
        label="ライトモードに切り替える"
        isActive={theme === "light"}
        onClick={() => setTheme("light")}
      >
        <FiSun className="w-4 h-4" />
      </ThemeToggleButton>
      <ThemeToggleButton
        label="ダークモードに切り替える"
        isActive={theme === "dark"}
        onClick={() => setTheme("dark")}
      >
        <FiMoon className="w-4 h-4" />
      </ThemeToggleButton>
    </div>
  );
}

type ThemeToggleButtonProps = {
  children: ReactNode;
  isActive: boolean;
  label: string;
  onClick: () => void;
};

function ThemeToggleButton({
  children,
  isActive,
  label,
  onClick,
}: ThemeToggleButtonProps) {
  return (
    <button
      type="button"
      onClick={onClick}
      className={clsx(
        BUTTON_BASE,
        isActive
          ? "text-white shadow-sm bg-primary-500"
          : "text-base-800 hover:bg-primary-50 hover:text-secondary-950 dark:text-base-100 dark:hover:bg-base-800 dark:hover:text-base-50"
      )}
      aria-pressed={isActive}
      aria-label={label}
    >
      {children}
    </button>
  );
}
