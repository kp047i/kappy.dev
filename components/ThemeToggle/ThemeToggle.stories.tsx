import { Meta, StoryObj } from "@storybook/nextjs";
import { expect, userEvent, within } from "storybook/test";

import { ThemeToggle } from "./ThemeToggle";

const meta: Meta<typeof ThemeToggle> = {
  title: "Components/ThemeToggle",
  component: ThemeToggle,
  parameters: {
    layout: "centered",
  },
};

export default meta;

type Story = StoryObj<typeof ThemeToggle>;

export const LightMode: Story = {
  render: () => <ThemeToggle />,
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const lightButton = await canvas.findByRole("button", {
      name: "ライトモードに切り替える",
    });
    const darkButton = await canvas.findByRole("button", {
      name: "ダークモードに切り替える",
    });

    await expect(lightButton).toHaveAttribute("aria-pressed", "true");

    await userEvent.click(darkButton);

    await expect(darkButton).toHaveAttribute("aria-pressed", "true");
    await expect(lightButton).toHaveAttribute("aria-pressed", "false");
  },
};

export const DarkMode: Story = {
  render: () => <ThemeToggle />,
  globals: {
    theme: "dark",
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const darkButton = await canvas.findByRole("button", {
      name: "ダークモードに切り替える",
    });
    await expect(darkButton).toHaveAttribute("aria-pressed", "true");
  },
};
