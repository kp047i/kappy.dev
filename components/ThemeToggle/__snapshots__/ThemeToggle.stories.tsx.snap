// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Components/ThemeToggle DarkMode play-test 1`] = `
<div class="flex items-center gap-1 rounded-lg border border-secondary-100/80 bg-overlay-100/90 p-1 shadow-sm backdrop-blur dark:border-base-800/60 dark:bg-base-800/60">
  <button type="button"
          class="flex h-9 w-9 items-center justify-center rounded-md border border-transparent text-sm transition-colors duration-200 text-base-800 hover:bg-primary-50 hover:text-secondary-950 dark:text-base-100 dark:hover:bg-base-800 dark:hover:text-base-50"
          aria-pressed="false"
          aria-label="ライトモードに切り替える"
  >
    <svg stroke="currentColor"
         fill="none"
         stroke-width="2"
         viewbox="0 0 24 24"
         stroke-linecap="round"
         stroke-linejoin="round"
         class="h-4 w-4"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <circle cx="12"
              cy="12"
              r="5"
      >
      </circle>
      <line x1="12"
            y1="1"
            x2="12"
            y2="3"
      >
      </line>
      <line x1="12"
            y1="21"
            x2="12"
            y2="23"
      >
      </line>
      <line x1="4.22"
            y1="4.22"
            x2="5.64"
            y2="5.64"
      >
      </line>
      <line x1="18.36"
            y1="18.36"
            x2="19.78"
            y2="19.78"
      >
      </line>
      <line x1="1"
            y1="12"
            x2="3"
            y2="12"
      >
      </line>
      <line x1="21"
            y1="12"
            x2="23"
            y2="12"
      >
      </line>
      <line x1="4.22"
            y1="19.78"
            x2="5.64"
            y2="18.36"
      >
      </line>
      <line x1="18.36"
            y1="5.64"
            x2="19.78"
            y2="4.22"
      >
      </line>
    </svg>
  </button>
  <button type="button"
          class="flex h-9 w-9 items-center justify-center rounded-md border border-transparent text-sm transition-colors duration-200 text-white shadow-sm bg-primary-500"
          aria-pressed="true"
          aria-label="ダークモードに切り替える"
  >
    <svg stroke="currentColor"
         fill="none"
         stroke-width="2"
         viewbox="0 0 24 24"
         stroke-linecap="round"
         stroke-linejoin="round"
         class="h-4 w-4"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z">
      </path>
    </svg>
  </button>
</div>
`;

exports[`Components/ThemeToggle LightMode play-test 1`] = `
<div class="flex items-center gap-1 rounded-lg border border-secondary-100/80 bg-overlay-100/90 p-1 shadow-sm backdrop-blur dark:border-base-800/60 dark:bg-base-800/60">
  <button type="button"
          class="flex h-9 w-9 items-center justify-center rounded-md border border-transparent text-sm transition-colors duration-200 text-white shadow-sm bg-primary-500"
          aria-pressed="true"
          aria-label="ライトモードに切り替える"
  >
    <svg stroke="currentColor"
         fill="none"
         stroke-width="2"
         viewbox="0 0 24 24"
         stroke-linecap="round"
         stroke-linejoin="round"
         class="h-4 w-4"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <circle cx="12"
              cy="12"
              r="5"
      >
      </circle>
      <line x1="12"
            y1="1"
            x2="12"
            y2="3"
      >
      </line>
      <line x1="12"
            y1="21"
            x2="12"
            y2="23"
      >
      </line>
      <line x1="4.22"
            y1="4.22"
            x2="5.64"
            y2="5.64"
      >
      </line>
      <line x1="18.36"
            y1="18.36"
            x2="19.78"
            y2="19.78"
      >
      </line>
      <line x1="1"
            y1="12"
            x2="3"
            y2="12"
      >
      </line>
      <line x1="21"
            y1="12"
            x2="23"
            y2="12"
      >
      </line>
      <line x1="4.22"
            y1="19.78"
            x2="5.64"
            y2="18.36"
      >
      </line>
      <line x1="18.36"
            y1="5.64"
            x2="19.78"
            y2="4.22"
      >
      </line>
    </svg>
  </button>
  <button type="button"
          class="flex h-9 w-9 items-center justify-center rounded-md border border-transparent text-sm transition-colors duration-200 text-base-800 hover:bg-primary-50 hover:text-secondary-950 dark:text-base-100 dark:hover:bg-base-800 dark:hover:text-base-50"
          aria-pressed="false"
          aria-label="ダークモードに切り替える"
  >
    <svg stroke="currentColor"
         fill="none"
         stroke-width="2"
         viewbox="0 0 24 24"
         stroke-linecap="round"
         stroke-linejoin="round"
         class="h-4 w-4"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z">
      </path>
    </svg>
  </button>
</div>
`;
